libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+            while ( n < 0)
+                  n += 12;
+            return n;
+      };
+      auto get_digit = [](int digit) {
+            // a modifier with no digit always references the 3rd
+            return (digit == FBIDigitNone ) ? 3 : digit;
+      };
+      auto render_harmony = [&] (FiguredBassItem::Modifier prefix, int relpitch, FiguredBassItem::Modifier suffix) {
+            relpitch = get_digit(relpitch);
+            if ( prefix != none && suffix != none) {
+                  qDebug("(1242) measure #%d, cannot render FiguredBassModifier with two modifiers", segment->measure()->no());
+                  return;
+            }
+            FiguredBassItem::Modifier modifier = (prefix == none) ? suffix : prefix;
+            ClefType clef = chord->staff()->clef(segment->tick());
+            
+            // calculate native pitch for line adjusted for transposing instrument.
+            // in doing so we have to SUBTRACT(relpitch-1) because lines are oriented from top to bottom, and
+            // pitches are oriented from bottom to top.  The 1 is because chord harmonies are noted as 1,3,5 (for a normal triad)
+            // rather than (0,2,4).
+            int harmony_line = note->line() - relpitch + 1;
+            int native_pitch = line2pitch(harmony_line, clef, Key::C) + note->ppitch() - note->epitch();
+
+            AccidentalVal acciv = segment->measure()->findAccidental(segment, chord->staff()->idx(), harmony_line);
+            int harmony;
+            switch ( modifier ) {
+                  case FiguredBassItem::Modifier::NONE: {
+                        Note *antecedent = nullptr;
+                        int antecedent_tick = -1; // if we fail to find an antecedence, this will remain -1.
+                        // Here we must apply the accidental seen earlier in the measure if there is one, even if in a different octave.
+                        // find the note with maximum tick, but <= note->tick() whose line % 7 == native_pitch->line%7, take its accidental.
+                        chord->score()->visitNotesInStaff(chord->measure(), [&](Note *n) {
+
+                              if ( n->track() != chord->track())
+                                    return;
+
+                              // we have to skip notes which are not the bottom note of the chord.
+                              // why? because these notes were auto-generated, and their note->line() returns 0 even if not really on line 0.
+                              if ( find_if(n->chord()->notes().begin(),
+                                           n->chord()->notes().end(),
+                                           [n](Note *j) {
+                                                 // if there is a note j whose pitch is < n's pitch, then n is not the bottom note, so return.
+                                                 return j->pitch() < n->pitch();
+                                           }) != n->chord()->notes().end())
```

Comment:
I wouldn’t get too upset about these issues at the moment.   As I said before, this was just a committed version of the code to get feedback of the behavior.
And yes, once the exact looping structure is/was settled, I don’t need the visitNotesInStaff function anyway.  I just had it there to
separate the two different questions of how to loop through notes, from what do do with notes once I’ve found them.

I’m quite sure the final code will look very different.

> On 24 May 2015, at 17:28, Marc Sabatella <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30953227>:
> 
> > +                        int antecedent_tick = -1; // if we fail to find an antecedence, this will remain -1.
> > +                        // Here we must apply the accidental seen earlier in the measure if there is one, even if in a different octave.
> > +                        // find the note with maximum tick, but <= note->tick() whose line % 7 == native_pitch->line%7, take its accidental.
> > +                        chord->score()->visitNotesInStaff(chord->measure(), [&](Note *n) {
> > +
> > +                              if ( n->track() != chord->track())
> > +                                    return;
> > +
> > +                              // we have to skip notes which are not the bottom note of the chord.
> > +                              // why? because these notes were auto-generated, and their note->line() returns 0 even if not really on line 0.
> > +                              if ( find_if(n->chord()->notes().begin(),
> > +                                           n->chord()->notes().end(),
> > +                                           [n](Note *j) {
> > +                                                 // if there is a note j whose pitch is < n's pitch, then n is not the bottom note, so return.
> > +                                                 return j->pitch() < n->pitch();
> > +                                           }) != n->chord()->notes().end())
> On Sun, May 24, 2015 at 3:55 AM, Maurizio M. Gavioli < notifications@github.com> wrote: It may be my aging brain, but for me it is extremely confusing the use of an inline function, moreover inside another inline function inside another inline function. At this point, I have lost any track of what is what.
> It's not just you. I believe can follow this, but it is extremely un-C++ like. I mean, obvious, C++ *supports* this syntax, but it doesn't *encourage* or *recommend* it for this purpose. I gather that sort of thing is common in Lisp, perhaps because it doesn't provide the same sort of procedural flow that C++ does, but there is no reason to resort to that sort of thing here. This seems like an ordinary loop performing an ordinary test, I see no no reason it shouldn't look like the those loops generally look elsewhere in the code.
> … <x-msg://5/#>
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30953227>.
>